File: scripts/publisher.py
#!/usr/bin/env python3
# encoding: utf-8

import sys
from typing import Final

import rospy

# custom message
from catkin_template.msg import Message

# constants
NODE_NAME: Final[str] = 'py_publisher'

PUB_QUEUE_SIZE: Final[int] = 10
PUB_RATE_HZ: Final[float] = 1.0
PUB_TOPIC: Final[str] = 'message'


def main() -> None:

  # initialize ROS Node
  rospy.init_node(NODE_NAME, sys.argv)

  # create publisher
  publisher = rospy.Publisher(PUB_TOPIC, Message, queue_size=PUB_QUEUE_SIZE)

  # create message
  msg = Message(data='py::publisher::message')

  # define publish rate
  pub_rate = rospy.Rate(PUB_RATE_HZ)

  while not rospy.is_shutdown():
    rospy.loginfo_once(f"Publishing to '{rospy.resolve_name(PUB_TOPIC)}' at {PUB_RATE_HZ} Hz")

    # publish message
    publisher.publish(msg)
    
    pub_rate.sleep()


if __name__ == '__main__':
    main()

File: src/nodes/publisher.cpp
#include <cstdint> // uint32_t

#include <ros/ros.h> // NodeHandle, Publisher, Rate

// custom message
#include <catkin_template/Message.h>

// constants
static constexpr auto kNodeName = "cpp_publisher";

static constexpr auto kPubTopic = "message";
static constexpr auto kPubRateHz = double{1.0};
static constexpr auto kPubQueueSize = std::uint32_t{10};


int main(int argc, char **argv) {

  // initialize ROS Node
  ros::init(argc, argv, kNodeName);

  auto handle = ros::NodeHandle();

  // create publisher
  ros::Publisher publisher = handle.advertise<catkin_template::Message>(kPubTopic, kPubQueueSize);

  // create message
  auto msg = catkin_template::Message();
  msg.data = "cpp::publisher::message";

  // define publish rate
  auto pub_rate = ros::Rate(kPubRateHz);

  while (ros::ok()) {
    ROS_INFO_ONCE("Publishing to '%s' at %f Hz", ros::names::resolve(kPubTopic).data(), kPubRateHz);

    // publish message
    publisher.publish(msg);

    ros::spinOnce();
    pub_rate.sleep();
  }

  return 0;
}